pydatastructs/trees/binary_trees.py

```diff
@@ -1053,3 +1054,132 @@ def get_sum(self, left_index, right_index):
+            elif self.tree[p].left == x and self.tree[self.tree[p].parent].right == p:
+                self._zag_zig(x, p)
+            else:
+                self._zig_zag(x, p)
+            p = self.tree[x].parent
+
+    def insert(self, key, x):
+        super(SelfBalancingBinaryTree, self).insert(key, x)
+        e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
+        self.tree[self.size-1].parent = p;
+        self.splay(e, p)
+
+    def delete(self, x):
+        e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
+        if e is None:
+            return
+        self.splay(e, p)
+        b = super(SelfBalancingBinaryTree, self).delete(x, balancing_info=True)
+        return True
+
+    def join(self, other):
+        """
+        Joins two trees self and other such that all elements of
+        the self tree are smaller than the elements of the other tree.
+
+        Parameters
+        ==========
+
+        other: SplayTree
+            SplayTree which needs to be joined with the self tree.
+
+        """
+        maxm = self.root_idx
+        while self.tree[maxm].right is not None:
+            maxm = self.tree[maxm].right
+        self.splay(maxm, self.tree[maxm].parent)
+        other.split_array(self, other.tree[other.root_idx].data)
+
+    def split(self, x):
+        """
+        Splits self tree into two trees such that one tree contains nodes
+        with key less than or equal to x and the other tree containing
+        nodes with key greater than x.
+
+        Parameters
+        ==========
+
+        x: key
+            Key of the element on the basis of which split is performed.
+
+        Returns
+        =======
+
+        other: SplayTree
+            SplayTree containing elements with key greater than x.
+
+        """
+        e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
+        if e is None:
+            return
+        self.splay(e, p)
+        other = SplayTree(None, None)
+        if self.tree[self.root_idx].right is not None:
+            self.split_array(other, self.tree[self.tree[self.root_idx].right].data)
+            self.tree[self.root_idx].right = None
+        return other
+
+    def split_array(self, sec, x):
+        """
+        Helper function that appends all elements of the SplayTree self into
+        SplayTree sec.
+
+        Parameters
+        ==========
+
+        sec: SplayTree
+            Elements of self tree are appended in sec.
+
+        x: key
+            Key of the node which needs to be appended in the sec tree.
+
+        """
+        super(SelfBalancingBinaryTree, sec).insert(x, x)
+        e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
+        if self.tree[e].left is not None:
+            l = self.tree[self.tree[e].left]
+            self.split_array(sec, l.data)
+        if self.tree[e].right is not None:
+            r = self.tree[self.tree[e].right]
+            self.split_array(sec, r.data)
```

Comment:
Can you convert this into iterative algorithm using stacks?

Reply:
Done.